net/connection.h

```diff
@@ -88,11 +89,18 @@ class Connection : public Threadable
      *
      * e.g. if it's an IMG command, it should be handled by Video::NetSource.
      */
-    void addCommandHandler(const std::string commandName, const CommandHandler handler)
+    void addCommandHandler(const std::string &commandName, const CommandHandler handler)
     {
+        std::lock_guard<std::mutex> guard(*m_CommandHandlersMutex);
         m_CommandHandlers.emplace(commandName, handler);
     }
 
+    void removeCommandHandler(const std::string &commandName)
+    {
+        std::lock_guard<std::mutex> guard(*m_CommandHandlersMutex);
+        m_CommandHandlers[commandName] = nullptr;
```

Comment:
Because at the moment if you send a command for which there isn't a handler you'll get a ``BadCommandError``. This wouldn't be the end of the world (as we're only removing the handlers when exiting anyway) but by setting them to null it means we can just silently ignore them until the ``Connection`` is destroyed.

Reply:
...but I agree that this isn't an obvious thing for it to do. Maybe we shouldn't use the word "remove" in the name...

Reply:
How about, we replace the add/remove functions with a single ``setCommandHandler`` function. Then when you wanted to add a handler you would do:
```
connection.setCommandHandler("TNK", driveMe);
```

And when you're finished you can do:
```
connection.setCommandHandler("TNK", nullptr);
```

The ``nullptr`` might make it more obvious that we're doing nothing for TNK command rather than removing it properly.

Reply:
I just think having a dictionary that can contain both empty slots **and** null pointers just in order to stop weird race conditions when you shut down is a bit gross. I think it would be better to either shut down in an order that prevents that happening or catch and ignore those exceptions

Reply:
I agree it isn't particularly pretty. I think the best option of the two would be try to shut down in the right order (ignoring incoming commands that don't have a handler means you won't realise if you've forgotten to add one), but unfortunately these objects are somewhat mutually dependent: e.g. you pass a ``Connection`` to ``Video::NetSource``'s constructor. So we would need to do something like this instead:
``
Video::NetSource vid;
Net::Connection connection = ....
vid.start(connection);
```
Maybe not a bad option. What do you think?

Reply:
*poke* :wink: